### ISSUE TITLE
Handle conflicts when copy and pasting

### ISSUE THREAD

**Issue by jelly** [cockpit-project/cockpit-files#467]
When we have a conflict when copy pasting after https://github.com/cockpit-project/cockpit-files/pull/457 we error out. That isn't what we want, we want to provide a user friendly way of making a choice as we do when uploading as can be seen below:

![image](https://github.com/cockpit-project/cockpit-files/assets/67428/8f4861ed-61bd-4147-902e-15fb59db5c29)

**Comment by briansmith0** [comment]
GNOME Files and KDE Dolphin also provide the opportunity to rename the file in this type of scenario.  

This other issue seems to be related to this as well:  https://github.com/cockpit-project/cockpit-files/issues/296  

I think it is a common workflow to go in to a directory, copy a file, and paste it in the same directory to make a backup of the file.

**Comment by jelly** [comment]
Testing this in GNOME files:

* copy paste in the same directory creates a new copy so "bar" becomes "bar (Copy)"
* paste into folder shows a conflict resolution dialog
![image](https://github.com/cockpit-project/cockpit-files/assets/67428/adab7a88-0811-4e9e-868e-3fca7d52d944)
* When changing directory and pasting you get the same behavior as "paste into".

**Comment by jelly** [comment]
It is not trivial to handle `cp --interactive` from Cockpit and for renaming we went with a Python script to run `renameat2`. So for copying we likely want to implement something ourselves as well some considerations:

* Permissions need to be retained (ACL/ SELinux?)
* Need to handle conflicts upfront
* Support renaming on conflict?
* Filesystem boundaries (relevant for moving)

For copying one option is:

https://docs.python.org/3/library/shutil.html#shutil.copytree

As passing `dirs_exist_ok` will throw an exception on an existing file it seems the only way we can handle resolutions well is overriding the `copy_function` . But this will be tricky to handle from the UI as it is callback based.



Some wild ideas:

* sendfile
* copyfile

### Moving

As moving is relevant I wonder if we can do this atomic by:

`mv foo /tmp/bar/foo`

```
fp = openat("/tmp/bar/foo")
renameat2("foo", fp);
```

Except when moving between two different filesystem, an easy test case:

```
source = b'/tmp/foo'
dest = '/foo'
>>> libc6.renameat2(AT_FDCWD, source, AT_FDCWD, dest, RENAME_NOREPLACE)
-1
>>> ctypes.get_errno()
18
>>> os.strerror(ctypes.get_errno())
'Invalid cross-device link'
```

**Comment by jelly** [comment]
Alternative suggestion from @allisonkarlitskaya is to not support pasting when the directory already exists. (As already implemented in https://github.com/cockpit-project/cockpit-files/pull/457

Only paste in directory is then harder to support but that is maybe a bit nieche.

**Comment by martinpitt** [comment]
> We allow copying multiple files at once and from different directories

Why do we do this? I don't know any file manager which supports this kind of "multiple clipboard". I see the use case for marking a bunch of files (or subdirs) in a directory view and moving/copying them someplace else -- but building a "additive persistent clipboard" with marking files from *different* dirs sounds like making our lives really hard..

**Comment by jelly** [comment]
> > We allow copying multiple files at once and from different directories
> 
> Why do we do this? I don't know any file manager which supports this kind of "multiple clipboard". I see the use case for marking a bunch of files (or subdirs) in a directory view and moving/copying them someplace else -- but building a "additive persistent clipboard" with marking files from _different_ dirs sounds like making our lives really hard..

I agree, it is valid to copy something from dir A and then browse to / and paste it but picking things up on the way is imo making life hard. So maybe the clipboard state should note be:

When in /tmp/foo and then copying "a, b" you'll get in the "clipboard":
 
"/tmp/foo/a" /tmp/foo/b"

Maybe it should be
```
{ files: [], dirs: [], base_dir: "/tmp/foo" }
```
And when copying in a different directory we compare $CWD with the `clipboard.base_dir` state and reset `files`, `dirs`.

This whole feature needs more thought.
